graph/patterns/viewpoint.md

```diff
@@ -0,0 +1,112 @@
+
+## Problem
+A shared resource, such as a website or a group message, may have different states for different users who access it at different times in an organizational context. For example, user1 may read and delete a message, while user2 may not have seen it yet. This usually happens when a shared item is presented in an individual context.
+## Solution
+
+The viewpoint pattern provides a solution to how to model an individual user context on a shared resource using a `viewpoint` structural property on an API entity type.
+For example, the `viewpoint` property can indicate whether a message is read, deleted, or flagged for a given user.
+The consistent naming convention ensures that when a developer uses Graph APIs all ‘viewpoints’ structural properties represent type specific user context across different M365 services and features.
+
+This pattern simplifies the API client logic by hiding the state transition details and providing state persistency on the server side. The server can manage the different viewpoints for the shared resource without exposing them to the client. You should also make this property filterable.
+## Issues and considerations
+
+- Because the `viewpoint` property reflects an individual user's context, it is null when accessed with application permissions.
+- Sometimes, the viewpoint can be computed on the server. In this case, an API producer should add OData annotations to the property to provide more information for downstream tools, such as SDKs and documentation generation.
+```
+    <Annotations Target="microsoft.graph.approvalItem/viewPoint">
+        <Annotation Term="Org.OData.Core.V1.Computed" Bool="true" ags:OwnerService="Microsoft.Approvals" />
+    </Annotations>
+```
+- An alternative to this design would be to store the user state on the client side. However, this may be problematic in some cases, because of the many devices that a user may have and the need to synchronize the state across them.
+
+## Examples
+
+### Defining a viewpoint
+
+The following example demonstrates how to define the 'viewpoint' property for the `chat` entity, where a chat is a collection of chatMessages between one or more participants: 
+```
+  <ComplexType Name="chatViewpoint" ags:WorkloadIds="Microsoft.Teams.GraphSvc">
+        <Property Name="isHidden" Type="Edm.Boolean" />
+        <Property Name="lastMessageReadDateTime" Type="Edm.DateTimeOffset" />
+  </ComplexType>
+
+  <EntityType Name="chat" BaseType="graph.entity" ags:OwnerService="Microsoft.Teams.GraphSvc">
+        <Property Name="chatType" Type="graph.chatType" Nullable="false" />
+        <Property Name="createdDateTime" Type="Edm.DateTimeOffset" />
+        <Property Name="lastUpdatedDateTime" Type="Edm.DateTimeOffset" />              
+        <Property Name="topic" Type="Edm.String" />
+        <Property Name="viewpoint" Type="graph.chatViewpoint" />
```

Comment:
As I understand this pattern, I have some concerns:
1. A URL locates a resource and the data in that resource should be agnostic to the client that is retrieving it.
2. I am also unclear how this sort of pattern would affect the 2-segment permissions scopes (e.g. if there were a `Chat.Read` segment, how would `viewpoint` be affected? Or the other way around, should `Chat.Read.All` give access to `viewpoint`, or should the client be required to have `Chat.Read.All` *and* `Chat.Read`.
3. This pattern also seems to encourage the use of delegate permissions with no corresponding app-only permission, which I don't think is a good idea. Let's say I want to write a script that let's a certain group of people know that I know they have messaged me but I've been busy. I want to read my unread chats, filter them by user, and then respond to the resulting chats. As a human, I don't want to be involved at all, but I can't get the necessary viewpoint with an app-only token. 

I would prefer a pattern that reverses the direction of the relationship. The `user` entity could have a property of `chatViews` which would be a collection of `chatViewpoint`s, which could have a navigation property to the `chat` entity that it is a view of. Doing this addresses all 3 of my concerns above because the data returned for a particular viewpoint is always relative to the user the viewpoint is on.

Reply:
Viewpoints are specific to a user; they don't exist for app permissions. This pattern reflects the current state where we have modeled multiple resources, such as Approvals, chats, and others, using this approach.

Reply:
I can align with everyone else if this is an established pattern and we are just writing it down for the first time, but I don't think I've ever heard this sort of pattern discussed.

Reply:
@corranrogue9 An alternative to the viewpoint design is to attach all the user specific state to a user instance.  However, there are cases where this cure is worse than the disease.  Having `/me/serviceAnnouncementStates` so that we can track which service announcements an admin has read would be a fairly horrible design to work with.

I need to double check, but my understanding is that viewpoints are always based on the identity of the calling user.  Not on a user mentioned in the path.  For this reason viewPoints should only be used on tenant relative entities.  Now I have to go find out what happens when you access the viewpoint of a chat that belongs to a user.